keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,407 @@
+ StartupPolicyOrder StartupPolicyOrderOptions `json:"startupPolicyOrder"`
+}
+```
+
+### Implementation
+
+We have three cases where startupPolicy applies:
+
+- when the jobset is first created
+- when the jobset is restarted after a failure (with restart or recreate all failure policy)
+- when the jobset is resumed after suspension
+
+In cases 1 and 2, the operator controls the order of creating the jobs, and so it creates the replicatedJob only when its turn comes in the startup sequence.
+In case 3, the operator controls the order of unsuspending the jobs, and so it unsuspends the replicatedJob only when its turn comes in the startup sequence.
+
+We use the annotation `jobset.sigs.k8s.io/replicated-job-to-start` to track which replicatedJob is next to create or unsuspend.
+
+### Defaulting/Validation
+
+- StartupPolicy is immutable.
+- StartupPolicyOrder of `Any` is default setting.
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+#### Unit Tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `controller`: `Aug 7th 2023` - `31.6%`
+
+Unit tests are low in this area as we do more of our testing at the integration level.
+
+We will add a new file called startup_policy to apply our actions.
+
+#### Integration tests
```

Comment:
Worth testing suspending a JobSet when it hasn't finished starting up.

Reply:
Good call. Added